File: analyzers/tests/SonarAnalyzer.Test/TestCases/ParameterNameMatchesOriginal.cs
using System;
using System.Collections.Generic;

namespace Tests.Diagnostics
{
    public partial class ParameterNamesInPartialMethod
    {
        partial void DoSomething(int x, int y);
        partial void DoSomething2(int x, int y);

        partial void DoSomething3(int x, int y);

        public void DoSomething4(int x, int y)
        {
        }
    }

    public partial class ParameterNamesInPartialMethod
    {
        partial void DoSomething(int x, int y)
        {

        }

        partial void DoSomething2(int someParam, int y) //Noncompliant {{Rename parameter 'someParam' to 'x' to match the partial class declaration.}}
//                                    ^^^^^^^^^
        {

        }

        partial void DoSomething3(int x, int y, int z);
    }

    public abstract class BaseClass
    {
        public virtual void DoSomethingVirtual(int x, int y)
        {
        }

        public abstract void DoSomethingAbstract(int x, int y);
    }

    public class ChildClass : BaseClass
    {
        public override void DoSomethingAbstract(int x, int someParam) //Noncompliant {{Rename parameter 'someParam' to 'y' to match the base class declaration.}}
        {
            throw new NotImplementedException();
        }

        public override void DoSomethingVirtual(int x, int someParam) //Noncompliant {{Rename parameter 'someParam' to 'y' to match the base class declaration.}}
        {
            base.DoSomethingVirtual(x, someParam);
        }
    }

    public abstract class ChildClassLevel2 : ChildClass
    {
        public override void DoSomethingAbstract(int x, int y) //Noncompliant {{Rename parameter 'y' to 'someParam' to match the base class declaration.}}
        {
            throw new NotImplementedException();
        }

        public override void DoSomethingVirtual(int x, int y) //Noncompliant {{Rename parameter 'y' to 'someParam' to match the base class declaration.}}
        {
            base.DoSomethingVirtual(x, y);
        }
    }

    public class InterfaceImplementation : IComparer<InterfaceImplementation>
    {
        public int Compare(InterfaceImplementation a, InterfaceImplementation y)
        {
            return 0;
        }
    }

    public partial interface IParameterNamesInPartialMethods
    {
        partial void DoSomething(int x, int y);
        partial void DoSomething2(int x, int y);
    }

    public partial interface IParameterNamesInPartialMethods
    {
        partial void DoSomething(int x, int y)
        {
        }

        partial void DoSomething2(int someParam, int y) //Noncompliant {{Rename parameter 'someParam' to 'x' to match the partial class declaration.}}
//                                    ^^^^^^^^^
        {
        }
    }

    public interface IGenericInterface<A>
    {
        void DoSomething();
        void DoSomething(A value);
        void DoSomething(A value, int intValue);
        void DoSomethingElse(A value);
        void DoSomethingElse(A value, ParameterClass parameterClassValue);
        void TryOneMoreTime(AnotherParameterClass value);
        void DoSomethingCaseSensitive(A value, int intValue);
    }
    public class ParameterClass { }
    public class AnotherParameterClass { }
    public class Implementation : IGenericInterface<ParameterClass>
    {
        public void DoSomething() { }
        public void DoSomething(ParameterClass parameter) { }
        public void DoSomething(AnotherParameterClass randomName) { }
        public void DoSomethingElse(ParameterClass completelyAnotherName) { }
        public void DoSomething(ParameterClass value, int myValue) { }                // Noncompliant
//                                                        ^^^^^^^
        public void DoSomethingElse(ParameterClass value, ParameterClass val) { }     // Noncompliant
//                                                                       ^^^
        public void TryOneMoreTime(AnotherParameterClass anotherParameter) { }        // Noncompliant
//                                                       ^^^^^^^^^^^^^^^^
        public void DoSomethingCaseSensitive(ParameterClass Value, int IntValue) { }  // Noncompliant
//                                                                     ^^^^^^^^
    }

    public struct StructImplementation : IGenericInterface<ParameterClass>
    {
        public void DoSomething() { }
        public void DoSomething(ParameterClass parameter) { }
        public void DoSomething(AnotherParameterClass randomName) { }
        public void DoSomethingElse(ParameterClass completelyAnotherName) { }
        public void DoSomething(ParameterClass value, int myValue) { }             // Noncompliant
//                                                        ^^^^^^^
        public void DoSomethingElse(ParameterClass value, ParameterClass val) { }  // Noncompliant
//                                                                       ^^^
        public void TryOneMoreTime(AnotherParameterClass anotherParameter) { }     // Noncompliant
//                                                       ^^^^^^^^^^^^^^^^
        public void DoSomethingCaseSensitive(ParameterClass Value, int IntValue) { }  // Noncompliant
//                                                                     ^^^^^^^^
    }

    public abstract class BaseClass<T>
    {
        public abstract void SomeMethod(T someParameter);

        public abstract void SomeMethod(T someParameter, int intParam);
    }

    public class ClassOne : BaseClass<int>
    {
        public override void SomeMethod(int renamedParam) { }

        public override void SomeMethod(int someParameter, int renamedParam) {  }  // Noncompliant
//                                                             ^^^^^^^^^^^^
    }

    public abstract class AbstractClassWithGenericMethod
    {
        abstract public void Foo<T>(T val);
        abstract public void Bar<T>(T val);
    }

    public class InheritedClassWithDefinition : AbstractClassWithGenericMethod
    {
        public override void Foo<T>(T myNewName) { }                               // Noncompliant
//                                    ^^^^^^^^^
        public override void Bar<T>(T val) { }
    }

    public interface IAnotherGenericInterface<A>
    {
        void DoSomething(A value);
        void DoSomething(A value, int intValue);
    }

    public interface IAnotherInterface : IAnotherGenericInterface<ParameterClass>
    {
        void DoSomethingElse(ParameterClass value);
    }

    public abstract class AnotherAbstractClass : IAnotherInterface
    {
        public abstract void DoSomething(ParameterClass abstractValue);
        public abstract void DoSomething(ParameterClass value, int IntValue);            //Noncompliant {{Rename parameter 'IntValue' to 'intValue' to match the interface declaration.}}
//                                                                 ^^^^^^^^
        public abstract void DoSomethingElse(ParameterClass Value);                      //Noncompliant {{Rename parameter 'Value' to 'value' to match the interface declaration.}}
//                                                          ^^^^^
    }

    public class AnotherImplementation : AnotherAbstractClass
    {
        public override void DoSomething(ParameterClass value) { }                       //Noncompliant {{Rename parameter 'value' to 'abstractValue' to match the base class declaration.}}
//                                                      ^^^^^
        public override void DoSomething(ParameterClass abstractValue, int IntValue) { } //Noncompliant {{Rename parameter 'abstractValue' to 'value' to match the base class declaration.}}
//                                                      ^^^^^^^^^^^^^
        public override void DoSomethingElse(ParameterClass Value) { }
    }

    // See https://github.com/SonarSource/sonar-dotnet/issues/4370
    public interface SomeInterface<A> { }

    public interface BaseInterface<A>
    {
        void Apply(SomeInterface<A> param);
    }

    public class BasicImplementation : BaseInterface<int>
    {
        public void Apply(SomeInterface<int> intValueParam) { }
    }

    public class StillGeneric<T> : BaseInterface<T>
    {
        public void Apply(SomeInterface<T> renamedParam) { }  // Noncompliant
    }

    public abstract class AbstractClass<T>
    {
        public abstract void Apply(SomeInterface<T> intValueParam);
    }

    public class OverridenCompliant : AbstractClass<int>
    {
        public override void Apply(SomeInterface<int> intValueParam) { }
    }

    public class OverridenNonCompliant<K> : AbstractClass<K>
    {
        public override void Apply(SomeInterface<K> renamedParam) { } // Noncompliant
    }

    public interface BaseInterface2<A, B>
    {
        void Apply(Tuple<A, B> param);
    }

    public class SomeClass<T> : BaseInterface2<T, int>
    {
        public void Apply(Tuple<T, int> renamed) { }
    }

    public abstract class BaseRepro
    {
        public abstract int SomeMethod(int SomeParam1, int SomeParam2);
    }

    public class Reproducer : BaseRepro
    {
        public override int SomeMethod(int _, int SomeParam2) => SomeParam2; // Noncompliant, it is not a discard
    }
}

File: analyzers/tests/SonarAnalyzer.Test/TestCases/ParameterNameMatchesOriginal.vb
'Partial methods are not relevant for VB.NET

Public MustInherit Class Base

    Public MustOverride Sub DoSomethingMustOverride(Name As String)

    Public Overridable Sub DoSomethingOverridable(Name As String)
    End Sub

    Public Sub NoParameters()
    End Sub

    Public Sub NoParametersNoParenthesis  'This should not have ()
    End Sub

End Class

Public Interface IContract

    Sub Add(Count As Integer)

End Interface

Public Class GoodUsage
    Inherits Base
    Implements IContract

    Public Overrides Sub DoSomethingMustOverride(Name As String)
    End Sub

    Public Overrides Sub DoSomethingOverridable(Name As String)
    End Sub

    Public Sub Add(Count As Integer) Implements IContract.Add
    End Sub

End Class

Public Class BadUsage
    Inherits Base
    Implements IContract

    Public Overrides Sub DoSomethingMustOverride(Description As String) ' Noncompliant {{Rename parameter 'Description' to 'Name' to match the base class declaration.}}
        '                                        ^^^^^^^^^^^
    End Sub

    Public Overrides Sub DoSomethingOverridable(Description As String)  ' Noncompliant
    End Sub

    Public Sub Add(Difference As Integer) Implements IContract.Add      ' Noncompliant {{Rename parameter 'Difference' to 'Count' to match the interface declaration.}}
        '          ^^^^^^^^^^
    End Sub

End Class

Public Class IgnoreCaseChange
    Implements IContract

    Public Sub Add(count As Integer) Implements IContract.Add   ' Compliant, differs only in casing
    End Sub

End Class

Public Interface IGenericInterface(Of A)

    Sub DoSomething()
    Sub DoSomething(Value As A)
    Sub DoSomething(Value As A, IntValue As Integer)
    Sub DoSomethingElse(Value As A)
    Sub DoSomethingElse(Value As A, ParameterClassValue As ParameterClass)
    Sub TryOneMoreTime(Value As AnotherParameterClass)
    Sub DoSomethingCaseSensitive(Value As A, IntValue As Integer)

End Interface

Public Class ParameterClass
End Class

Public Class AnotherParameterClass
End Class

Public Class Implementation
    Implements IGenericInterface(Of ParameterClass)

    Public Sub DoSomething() Implements IGenericInterface(Of ParameterClass).DoSomething
    End Sub

    Public Sub DoSomething(Parameter As ParameterClass) Implements IGenericInterface(Of ParameterClass).DoSomething
    End Sub

    Public Sub DoSomething(RandomName As AnotherParameterClass)
    End Sub

    Public Sub DoSomethingElse(CompletelyAnotherName As ParameterClass) Implements IGenericInterface(Of ParameterClass).DoSomethingElse
    End Sub

    Public Sub DoSomething(Value As ParameterClass, MyValue As Integer) Implements IGenericInterface(Of ParameterClass).DoSomething                             ' Noncompliant
        '                                           ^^^^^^^
    End Sub

    Public Sub DoSomethingElse(Value As ParameterClass, Val As ParameterClass) Implements IGenericInterface(Of ParameterClass).DoSomethingElse                  ' Noncompliant
        '                                               ^^^
    End Sub

    Public Sub TryOneMoreTime(AnotherParameter As AnotherParameterClass) Implements IGenericInterface(Of ParameterClass).TryOneMoreTime                               ' Noncompliant
        '                     ^^^^^^^^^^^^^^^^
    End Sub

    Public Sub DoSomethingCaseSensitive(VALUE As ParameterClass, INTVALUE As Integer) Implements IGenericInterface(Of ParameterClass).DoSomethingCaseSensitive
    End Sub

End Class

Public Structure StructImplementation
    Implements IGenericInterface(Of ParameterClass)

    Public Sub DoSomething() Implements IGenericInterface(Of ParameterClass).DoSomething
    End Sub

    Public Sub DoSomething(Parameter As ParameterClass) Implements IGenericInterface(Of ParameterClass).DoSomething
    End Sub

    Public Sub DoSomething(RandomName As AnotherParameterClass)
    End Sub

    Public Sub DoSomethingElse(CompletelyAnotherName As ParameterClass) Implements IGenericInterface(Of ParameterClass).DoSomethingElse
    End Sub

    Public Sub DoSomething(Value As ParameterClass, MyValue As Integer) Implements IGenericInterface(Of ParameterClass).DoSomething                             ' Noncompliant
        '                                           ^^^^^^^
    End Sub

    Public Sub DoSomethingElse(Value As ParameterClass, Val As ParameterClass) Implements IGenericInterface(Of ParameterClass).DoSomethingElse                  ' Noncompliant
        '                                               ^^^
    End Sub

    Public Sub TryOneMoreTime(AnotherParameter As AnotherParameterClass) Implements IGenericInterface(Of ParameterClass).TryOneMoreTime                               ' Noncompliant
        '                     ^^^^^^^^^^^^^^^^
    End Sub

    Public Sub DoSomethingCaseSensitive(VALUE As ParameterClass, INTVALUE As Integer) Implements IGenericInterface(Of ParameterClass).DoSomethingCaseSensitive
    End Sub

End Structure

Public MustInherit Class BaseClass(Of T)

    Public MustOverride Sub SomeMethod(SomeParameter As T)
    Public MustOverride Sub SomeMethod(SomeParameter As T, IntParam As Integer)

End Class

Public Class ClassOne
    Inherits BaseClass(Of Integer)

    Public Overrides Overloads Sub SomeMethod(RenamedParam As Integer)
    End Sub

    Public Overrides Overloads Sub SomeMethod(SomeParameter As Integer, RenamedParam As Integer)  ' Noncompliant
        '                                                               ^^^^^^^^^^^^
    End Sub

End Class

Public MustInherit Class AbstractClassWithGenericMethod

    MustOverride Public Sub Foo(Of T)(Val As T)
    MustOverride Public Sub Bar(Of T)(Val As T)

End Class

Public Class InheritedClassWithDefinition
    Inherits AbstractClassWithGenericMethod

    Public Overrides Sub Foo(Of T)(MyNewName As T)  ' Noncompliant
        '                          ^^^^^^^^^
    End Sub

    Public Overrides Sub Bar(Of T)(Val As T)
    End Sub

End Class

Public Interface IAnotherGenericInterface(Of A)

    Sub DoSomething(Value As A)
    Sub DoSomething(Value As A, IntValue As Integer)

End Interface

Public Interface IAnotherInterface
    Inherits IAnotherGenericInterface(Of ParameterClass)

    Sub DoSomethingElse(Value As ParameterClass)

End Interface

Public MustInherit Class AnotherAbstractClass
    Implements IAnotherInterface
    Public MustOverride Sub DoSomething(AbstractValue As ParameterClass) Implements IAnotherGenericInterface(Of ParameterClass).DoSomething
    Public MustOverride Sub DoSomething(Value As ParameterClass, IntValue As Integer) Implements IAnotherGenericInterface(Of ParameterClass).DoSomething
    Public MustOverride Sub DoSomethingElse(Value As ParameterClass) Implements IAnotherInterface.DoSomethingElse

End Class

Public Class AnotherImplementation
    Inherits AnotherAbstractClass

    Public Overrides Overloads Sub DoSomething(Value As ParameterClass)  'Noncompliant {{Rename parameter 'Value' to 'AbstractValue' to match the base class declaration.}}
        '                                      ^^^^^
    End Sub

    Public Overrides Overloads Sub DoSomething(AbstractValue As ParameterClass, IntValue As Integer)  'Noncompliant {{Rename parameter 'AbstractValue' to 'Value' to match the base class declaration.}}
        '                                      ^^^^^^^^^^^^^
    End Sub

    Public Overrides Sub DoSomethingElse(Value As ParameterClass)
    End Sub

End Class

' See https://github.com/SonarSource/sonar-dotnet/issues/4370
Public Interface SomeInterface(Of A)
End Interface

Public Interface BaseInterface(Of A)
    Sub Apply(ByVal param As SomeInterface(Of A))
End Interface

Public Class BasicImplementation
    Implements BaseInterface(Of Integer)

    Public Sub Apply(ByVal intValueParam As SomeInterface(Of Integer)) Implements BaseInterface(Of Integer).Apply
    End Sub
End Class

Public Class StillGeneric(Of T)
    Implements BaseInterface(Of T)

    Public Sub Apply(ByVal renamedParam As SomeInterface(Of T)) Implements BaseInterface(Of T).Apply  ' Noncompliant
    End Sub
End Class

Public MustInherit Class AbstractClass(Of T)
    Public MustOverride Sub Apply(ByVal param As SomeInterface(Of T))
End Class

Public Class OverridenCompliant
    Inherits AbstractClass(Of Integer)

    Public Overrides Overloads Sub Apply(ByVal intValueParam As SomeInterface(Of Integer))
    End Sub
End Class

Public Class OverridenNonCompliant(Of K)
    Inherits AbstractClass(Of K)

    Public Overrides Overloads Sub Apply(ByVal renamedParam As SomeInterface(Of K)) ' Noncompliant
    End Sub
End Class
